protocol/shannon/fullnode_cache.go

```diff
@@ -161,22 +215,111 @@ func (cfn *cachingFullNode) GetSession(
 	serviceID protocol.ServiceID,
 	appAddr string,
 ) (sessiontypes.Session, error) {
+
+	height, err := cfn.GetCurrentBlockHeight(ctx)
+	if err != nil {
+		return sessiontypes.Session{}, err
+	}
+	sessionKey := getSessionCacheKey(serviceID, appAddr, height)
+
 	// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#get-or-fetch
 	return cfn.sessionCache.GetOrFetch(
 		ctx,
-		getSessionCacheKey(serviceID, appAddr),
+		sessionKey,
 		func(fetchCtx context.Context) (sessiontypes.Session, error) {
-			cfn.logger.Debug().Str("session_key", getSessionCacheKey(serviceID, appAddr)).Msgf(
+			cfn.logger.Debug().Str("session_key", sessionKey).Msgf(
 				"[cachingFullNode.GetSession] Fetching from full node",
 			)
 			return cfn.lazyFullNode.GetSession(fetchCtx, serviceID, appAddr)
 		},
 	)
 }
 
-// getSessionCacheKey builds a unique cache key for session: <prefix>:<serviceID>:<appAddr>
-func getSessionCacheKey(serviceID protocol.ServiceID, appAddr string) string {
-	return fmt.Sprintf("%s:%s:%s", sessionCacheKeyPrefix, serviceID, appAddr)
+// GetSessionWithGracePeriod implements grace period aware session fetching with caching.
```

Comment:
#PUC Current comment doesn't explain WHY this method exists.

Would be great to clarify:
- The distributed coordination problem it solves
- When it returns previous vs current session
- How it prevents request failures during rollovers